kernel/src/schema/diff.rs

```diff
@@ -0,0 +1,2549 @@
+        | FieldChangeType::NullabilityTightened
+        | FieldChangeType::ContainerNullabilityTightened
+        | FieldChangeType::MetadataChanged => true,
+        FieldChangeType::Multiple(multiple_changes) => {
+            multiple_changes.iter().any(is_breaking_change_type)
+        }
+        _ => false,
+    }
+}
+
+/// Computes whether the diff contains breaking changes
+fn compute_has_breaking_changes(
+    added_fields: &[FieldChange],
+    _removed_fields: &[FieldChange],
+    updated_fields: &[FieldUpdate],
+) -> bool {
+    // Removed fields are safe - existing data files remain valid, queries referencing
+    // removed fields will fail at query time but data integrity is maintained
+    // Adding a non-nullable (required) field is breaking - existing data won't have values
+    added_fields.iter().any(|add| !add.field.nullable)
+        // Certain update types are breaking (type changes, nullability tightening, etc.)
+        || updated_fields
+            .iter()
+            .any(|update| is_breaking_change_type(&update.change_type))
+}
+
+/// Filters field changes to keep only the least common ancestors (LCA).
+///
+/// This filters out descendant fields when their parent is also in the set.
+/// For example, if both "user" and "user.name" are in the input, this returns only "user"
+/// since reporting "user.name" would be redundant.
+///
+/// The algorithm is O(n) where n is the number of fields:
+/// 1. Put all paths in a HashSet for O(1) lookup
+/// 2. For each field, check if its immediate parent is in the set
+/// 3. Keep only fields whose parent is NOT in the set
+fn filter_ancestor_fields(fields: Vec<FieldChange>) -> Vec<FieldChange> {
+    // Build a set of all paths for O(1) lookup (owned to avoid lifetime issues)
+    let all_paths: HashSet<ColumnName> = fields.iter().map(|f| f.path.clone()).collect();
+
+    // Filter to keep only fields whose parent is NOT in the set
+    fields
+        .into_iter()
+        .filter(|field_change| {
+            let path_parts = field_change.path.path();
+
+            // Top-level fields (length 1) have no parent, so keep them
+            if path_parts.len() == 1 {
+                return true;
+            }
+
+            // Construct parent path by removing the last component
+            let parent_path = ColumnName::new(&path_parts[..path_parts.len() - 1]);
+
+            // Keep this field only if its parent was NOT in the input set
+            !all_paths.contains(&parent_path)
+        })
+        .collect()
+}
+
+/// Checks if a field path has a parent in the given set of ancestor paths.
+///
+/// Returns true if any path in `added_ancestor_paths` is a prefix of `path`.
+/// For example, "user" is an ancestor of "user.name" and "user.address.street".
+fn has_added_ancestor(path: &ColumnName, added_ancestor_paths: &HashSet<ColumnName>) -> bool {
+    added_ancestor_paths
+        .iter()
+        .any(|ancestor| path.is_descendant_of(ancestor))
```

Comment:
Hmm this iterates over all the ancestor paths. That's O(N) for N added columns. 

But we can actually check each path prefix:
```rust
    let parent_opt = path.parent();
    while let Some(curr) = parent_opt {
        if added_ancestor_paths.contains(curr) {
            return true;
        }
        parent_opt = curr.parent()
    }
```